### PR TITLE
Vectorized CPU code implementing right shift operator.

### PR DESCRIPTION
**Pull request by alexsamardzic**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89284
* __->__ #88990

cc @VitalyFedyunin @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/cpu/vec/vec256/vec256_int.h [hunk 1]
```diff
@@ -1328,6 +1340,38 @@ Vectorized<int8_t> inline operator<<(const Vectorized<int8_t>& a, const Vectoriz
...
+  // Shift the number logically to the right, thus filling the most
+  // significant bits with 0s.  Then, replace these bits with the sign
+  // bit.
+  __m256i sign_bits = _mm256_cmpgt_epi64(_mm256_set1_epi64x(0), a);
+  __m256i b_inv_mod_64 = _mm256_sub_epi64(_mm256_set1_epi64x(64), b);
+  __m256i sign_ext = _mm256_sllv_epi64(sign_bits, b_inv_mod_64);
+  __m256i c = _mm256_srlv_epi64(a, b);
+  c = _mm256_or_si256(c, sign_ext);
+
+  return c;
```
Comment (sanchitintel) [inline aten/src/ATen/cpu/vec/vec256/vec256_int.h:1357]: A digression: @jgong5, do you think there's some way we can figure out whether or not 256-bit equivalents (which are not supported by AVX2, though) of AVX512 intrinsics result in as much downclocking as their 512-bit equivalents? Not sure if we'd have to check on a case-by-case basis, or if there's an established trend. 

For example, in this case, if `_mm256_srav_epi64` doesn't cause as much downclocking as  `_mm512_srav_epi64`, then we can add some changes in PyTorch to ensure that if a system supports AVX512 but `ATEN_CPU_CAPABILITY` is `avx2`, then we would still be able to leverage AVX512 intrinsics such as [`_mm256_srav_epi64`](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=_mm256_srav_epi64&ig_expand=6730,6733,6730).

Thanks!

Comment (jgong5) [reply aten/src/ATen/cpu/vec/vec256/vec256_int.h:1357]: Even though the 256 bit-width AVX512VL doesn't incur as much downclocking as 512 bit-width AVX512 but I don't think it is worth the complexity to create a new dispatching path here due to following reasons:
1. The downclocking from 512 bit-width AVX512 is not that obvious on new CPU generations. The benefit of 512 bit-width definitively outweighs the downclocking side-effect.
2. Different SKUs might have different number of AVX256 and AVX512 ports, plus the capability supporting arithmetic operations on different ports. Even if the number of AVX256 ports is bigger than AVX512 ones, the dispatching logic would be uarch and instruction specific, which would be fairly complicated.

File: aten/src/ATen/cpu/vec/vec_base.h [hunk 2]
```diff
@@ -839,6 +847,12 @@ inline Vectorized<T>& operator <<= (Vectorized<T>& a, const Vectorized<T>& b) {
   return a;
 }
 
+template <typename T>
+inline Vectorized<T>& operator >>= (Vectorized<T>& a, const Vectorized<T>& b) {
+  a = a >> b;
+  return a;
+}
+
```
Comment (lezcano) [inline aten/src/ATen/cpu/vec/vec_base.h:855]: I see that all the functions use this pattern. It's weird to me that we don't use the usual C++ pattern of implementing the out-of-place op in terms of the in-place one.
@peterbell10 any idea why do we do this?

**Review by lezcano — APPROVED** [review]
Not much to do left after your last PR. Let's wait to see if @peterbell10 has anything to add though.

**Review by peterbell10 — APPROVED** [review]

File: aten/src/ATen/cpu/vec/vec_base.h (hunk 2 above)
Comment (peterbell10) [reply aten/src/ATen/cpu/vec/vec_base.h:855]: Probably because the intrinsic operations are naturally out-of-place.

**Comment by sanchitintel** [comment]
Trunk seems to be broken for `test_jit_cuda_fuser` for the failed CI check:

image

**Pull request closed without merging**